sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorDispatcher.java

```diff
@@ -109,26 +115,35 @@ private void signalWorkQueue() throws IOException {
     private final class WorkScheduler implements Callback {
         @Override
         public void run(Selectable selectable) {
+            if (dequeueInProgress.getAndSet(true)) {
+                return;
+            }
+
             try {
-                ByteBuffer oneKbByteBuffer = ByteBuffer.allocate(1024);
-                while (ioSignal.source().read(oneKbByteBuffer) > 0) {
-                    // read until the end of the stream
-                    oneKbByteBuffer = ByteBuffer.allocate(1024);
+                try {
+                    ByteBuffer oneKbByteBuffer = ByteBuffer.allocate(1024);
+                    while (ioSignal.source().read(oneKbByteBuffer) > 0) {
+                        // read until the end of the stream
+                        oneKbByteBuffer = ByteBuffer.allocate(1024);
+                    }
+                } catch (ClosedChannelException ignorePipeClosedDuringReactorShutdown) {
+                    logger.info("WorkScheduler.run() failed with an error. Can be ignored.",
+                        ignorePipeClosedDuringReactorShutdown);
+                } catch (IOException ioException) {
+                    throw logger.logExceptionAsError(new RuntimeException(
+                        String.format("WorkScheduler.run() failed with an error: %s", ioException), ioException));
                 }
-            } catch (ClosedChannelException ignorePipeClosedDuringReactorShutdown) {
-                logger.info("WorkScheduler.run() failed with an error: %s", ignorePipeClosedDuringReactorShutdown);
-            } catch (IOException ioException) {
-                logger.error("WorkScheduler.run() failed with an error: %s", ioException);
-                throw logger.logExceptionAsError(new RuntimeException(ioException));
-            }
 
-            Work topWork;
-            while ((topWork = workQueue.poll()) != null) {
-                if (topWork.delay != null) {
-                    reactor.schedule((int) topWork.delay.toMillis(), topWork.dispatchHandler);
-                } else {
-                    topWork.dispatchHandler.onTimerTask(null);
+                Work topWork;
+                while ((topWork = workQueue.poll()) != null) {
+                    if (topWork.delay != null) {
+                        reactor.schedule((int) topWork.delay.toMillis(), topWork.dispatchHandler);
+                    } else {
+                        topWork.dispatchHandler.onTimerTask(null);
+                    }
                 }
+            } finally {
+                dequeueInProgress.set(false);
```

Comment:
There is still a race condition here if a thread has reached this line but hasn't executed it and concurrently, another thread checks the status (still `true` as this line is not executed yet) and returns thinking that dequeue is in progress whereas the dequeue has ended and no thread is actually dequeuing anything.

Reply:
I feel like I'm missing something because 118 checks using getAndSet and returns from the method if there was something in the loop already. I updated it to compareAndSet so we explicitly check for false and set it to true, otherwise return from method. Does that help?
https://github.com/Azure/azure-sdk-for-java/pull/20221/files#diff-5a4c901a968275701657666e85e1eff354ee15872b58bfad64003f4c1c7c05c9R118